### PR TITLE
Volume coupling

### PR DESCRIPTION
**Pull request by tirgendetwas**
This is a solution to issue #229 .

Volume coupling implementation for the OpenFOAM adapter. Our main goals are:

**1. The coupling is integrated with existing features** (currently planned for FF and CHT modules):
We started building upon PR #183 but have since decided against a separate Volume Coupling module. The volume coupling is being implemented for each state variable (each variable which would make sense to couple over a volume).

**2. Configurable coupling region:**
Implemented in [PR#270](https://github.com/precice/openfoam-adapter/pull/270)

Relevant tutorials: [channel transport](https://github.com/precice/tutorials/pull/315) and [volume-coupled flow](https://github.com/precice/tutorials/pull/350)

TODO list:

- [x] I updated the documentation in `docs/`
- [x] I added a changelog entry in `changelog-entries/` (create directory if missing)

### CODE REVIEW

**Comment by LouieVoit** [comment]
Hi,

Don't we want to be able to use the volume coupling also for CHT, FF or FSI ? I'm not sure I understand why it's a separate module. 
Also, I've extended the adapter to use volume coupling in my own fork of the adapter. Within the Interface.C file, I've added a zone option in the preciceDict dictionary to define OpenFOAM zones to use for the coupling. That way one does not have to use all the mesh for the coupling. Should it be added to this PR ?

Thanks :-)
Louis

**Comment by MakisH** [comment]
Hi @LouieVoit,

we started from a past contribution that already implemented volume coupling as a separate module (#183). We have already discussed and concluded that this is not the right approach. We also wanted to add zones, similarly to #84.

This should eventually become clear in the description of this PR (todo @tirgendetwas).

File: Interface.C
Comment (MakisH) [inline Interface.C:297]:
```suggestion
        // and add those to the previously determined number of mesh points in the volume
```

File: Interface.C
Comment (MakisH) [inline Interface.C:293]:
```suggestion
        // Get the number of (volume centered) mesh points in the volume
```

File: Interface.C
Comment (MakisH) [inline Interface.C:319]:
```suggestion
        for (uint i = 0; i < CellCenters.size(); i++)
```

File: FF/Velocity.C
Comment (MakisH) [inline FF/Velocity.C:115]: This is about supporting mesh connectivity. Do we support mesh connectivity for volume meshes?

File: CouplingDataUser.H
Comment (MakisH) [inline CouplingDataUser.H:17]: Would it make more sense to call this `cellCenters`, as I think you have it elsewhere? I am not sure.

File: CHT/Temperature.C
Comment (MakisH) [inline CHT/Temperature.C:107]: Same here: specific to connectivity.

**Review by MakisH — COMMENTED** [review]
Some first comments (not tested yet).

File: CHT/HeatFlux.C
Comment (tirgendetwas) [reply CHT/HeatFlux.C:40]: `.clang-format` actually isn't an empty file, but for some reason the git changelog lists it that way. It seems to be the same as the adapter's original `.clang-format`, I'm not sure what happened in [this commit](https://github.com/precice/openfoam-adapter/commit/579c5e4c511437009a61950baaafbc5b91ce5bae) 

Also, installing `clang-format 11.0.1` and applying the script again made no difference.

Something is not completely right...

File: CouplingDataUser.H
Comment (tirgendetwas) [reply CouplingDataUser.H:17]: Maybe something like `volumeCenters`? I think it would make sense to keep "volume" in the name, maybe combine the two ideas :smile:

File: CHT/HeatFlux.C
Comment (MakisH) [reply CHT/HeatFlux.C:40]: Strange... I will have a look before merging, let's keep this unresolved.

File: CouplingDataUser.H
Comment (MakisH) [reply CouplingDataUser.H:17]: I like the name! Go ahead.

File: FF/Velocity.C
Comment (MakisH) [reply FF/Velocity.C:115]: What I meant in this comment was not to implement mesh connectivity as part of this PR (even though it would be nice, but let's document it in an issue first). I just meant to remove this addition, because we don't support it yet.

Comment (MakisH) [reply FF/Velocity.C]: Your original suggestion was in the right direction. I got confused, as the connectivity feature was added on top of everything much later, but the architecture did not really allow us for a nice implementation.

If we add the `|| this->locationType_ == LocationType::volume`, this means that volume centers are supported regardless of mesh connectivity or not. However, we need to distinguish here between mesh connectivity (supporting only `faceCenters`) and no mesh connectivity (supporting both).

As you already pointed out in a direct message, we can do this similarly to the Temperature class.

File: FF/Pressure.C
Comment (MakisH) [inline FF/Pressure.C:21]: What does this comment refer to?
Would a comment consistent with the "For every boundary patch of the interface" (i.e., "For every internal cell of the region") be more helpful to understand the intention?

File: CHT/Temperature.C
Comment (MakisH) [inline CHT/Temperature.C:103]: Would a positive formulation ("if faceCenters or faceNodes") be clearer, safer, and consistent, e.g., with the `Pressure.C` and `Velocity.C`?
The comment can remain.

File: Interface.C
Comment (MakisH) [inline Interface.C:283]:
```suggestion
```

File: Interface.C
Comment (MakisH) [inline Interface.C:288]: I am not sure I understand this comment. Do you mean that the branch that we started from did not yet have `faceNodes` implemented or complete?

In any case, we are assuming cell-based FVM, which is the most common approach (there is also nodes-based FVM, but this is way out of scope of the complete adapter at the moment, as only some extensions of OpenFOAM implement it). As such, all solvers store all fields (with some exceptions) in the centers of the cells.

Independent on what the development was based upon, why does it matter?

File: Interface.C
Comment (MakisH) [inline Interface.C:125]:
```suggestion
```

File: Interface.C
Comment (MakisH) [inline Interface.C:291]: So, does this identify all points in the volume region + all points on the face centers of the region? Why do we need that?

I would expect that, when coupling a volume, we don't couple the boundaries. Or at least not with the same values. I would expect that the boundaries of the volume are specified by the solver, based on the rest of the domain. This makes mostly sense when coupling only one region of the whole mesh.

**Review by MakisH — COMMENTED** [review]
Some comments, as I am trying to get an overview of the current state, and keep the rock rolling. No rush to implement.

The main aspect that you may already want to check, from a debugging perspective, is how you specify the locations to couple (volume + patches, while I think it should only be volume).

File: Interface.C
Comment (MakisH) [inline Interface.C:319]: I am getting a compiler warning with OpenFOAM v2212 and GCC 11.3:

```
g++ -std=c++14 -m64 -pthread -DOPENFOAM=2212 -DWM_DP -DWM_LABEL_SIZE=32 -Wall -Wextra -Wold-style-cast -Wnon-virtual-dtor -Wno-unused-parameter -Wno-invalid-offsetof -Wno-attributes -Wno-unknown-pragmas -O3  -DNoRepository -ftemplate-depth-100 -fdiagnostics-color=always -I/usr/lib/openfoam/openfoam2212/src/finiteVolume/lnInclude -I/usr/lib/openfoam/openfoam2212/src/meshTools/lnInclude -I/usr/lib/openfoam/openfoam2212/src/transportModels/ -I/usr/lib/openfoam/openfoam2212/src/transportModels/incompressible/lnInclude -I/usr/lib/openfoam/openfoam2212/src/transportModels/compressible/lnInclude -I/usr/lib/openfoam/openfoam2212/src/transportModels/twoPhaseMixture/lnInclude -I/usr/lib/openfoam/openfoam2212/src/transportModels/interfaceProperties/lnInclude -I/usr/lib/openfoam/openfoam2212/src/transportModels/immiscibleIncompressibleTwoPhaseMixture/lnInclude -I/usr/lib/openfoam/openfoam2212/src/thermophysicalModels/basic/lnInclude -I/usr/lib/openfoam/openfoam2212/src/TurbulenceModels/turbulenceModels/lnInclude -I/usr/lib/openfoam/openfoam2212/src/TurbulenceModels/compressible/lnInclude -I/usr/lib/openfoam/openfoam2212/src/TurbulenceModels/incompressible/lnInclude -I/usr/lib/openfoam/openfoam2212/src/triSurface/lnInclude -I/home/gc/repos/precice/precice/src -I../  -iquote. -IlnInclude -I/usr/lib/openfoam/openfoam2212/src/OpenFOAM/lnInclude -I/usr/lib/openfoam/openfoam2212/src/OSspecific/POSIX/lnInclude   -fPIC -c FSI/FSI.C -o Make/linux64GccDPInt32Opt/FSI/FSI.o
Interface.C: In member function ‘void preciceAdapter::Interface::configureMesh(const Foam::fvMesh&, const string&, const string&)’:
Interface.C:362:28: warning: comparison of integer expressions of different signedness: ‘uint’ {aka ‘unsigned int’} and ‘Foam::label’ {aka ‘int’} [-Wsign-compare]
  362 |         for (uint i = 0; i < CellCenters.size(); i++)
      |            
```

(not sure why I get different line numbers)

Comment (tirgendetwas) [reply Interface.C]: It seems that `CellCenters.size()` is of type `Foam::label` which is an `int64_t`/`int32_t` (signed) under the hood. Solved the problem by casting for now...

File: FF/Pressure.C
Comment (tirgendetwas) [reply FF/Pressure.C]: I added this comment for myself at an early stage and forgot to remove it later on. Fixed!

File: Interface.C
Comment (MakisH) [reply Interface.C]: Why not just `auto`? We can use C++11 features.

File: Interface.C
Comment (tirgendetwas) [reply Interface.C:286]: This is a comment from Prasad's original fork. It might have been relevant at the time, but if it isn't anymore I can remove it (or only leave the first line)

File: Interface.C
Comment (tirgendetwas) [reply Interface.C:289]: The coupling includes the cell centers on the whole domain (at this implementation stage) and the face centers for all specified patches - basically adding the cell centers of the volume to the `face nodes`-surface coupling case. 

So in a tutorial case which uses `locations    volumeCenters`, all of the internal fields will be coupled (cell centers), and additionally specified patches can be coupled as well (on face centers).

File: Interface.C
Comment (tirgendetwas) [reply Interface.C]: Thanks, that worked!

File: Interface.C
Comment (MakisH) [reply Interface.C:329]: I think we can remove this.

File: Interface.C
Comment (MakisH) [reply Interface.C:332]: So, if I provide `locations volumeCenters` and don't specify any patches, only the volume region will be modified, without the patches. If I also specify patches, these will also be modified. This sounds reasonable behavior to me, we just need to document it.

Do we need to use different coupling meshes for volumes and patches? I am trying to think of mapping issues, especially now that we have volume mapping as well. @uekerman

Comment (uekerman) [reply Interface.C:332]: mmh, not too sure about this (if I understood it correctly). Mixing surface and volume coupling here could confuse more than it helps. There could also be cases where we want to read values in the volume from one solver and additionally read values on the surface from another solver. I would always treat volume and surface as two different meshes in preCICE.
Or is there a specific use case where we would need this feature?

Comment (tirgendetwas) [reply Interface.C:328]: So how should we resolve this? Previous applications in that area have used both patches and volume also ([PR#88](https://github.com/precice/openfoam-adapter/pull/88/files) and [PR#97](https://github.com/precice/openfoam-adapter/pull/97/files)), so I suspected it might have been necessary and kept it like that. The channel transport [tutorial](https://github.com/precice/tutorials/pull/315) case I have set up looks very similar whether I couple patches + volume or only volume.

**Comment by mosayebshams** [comment]
Hi @MakisH and @tirgendetwas,

I was wondering if there was a toy-case for the volume-coupling to guide me through adjusting `precice-config.xml` and `preciceDict` files to read pressure data from specific volume locations in a fluid-openfoam simulation.

File: docs/README.md
Comment (MakisH) [inline docs/README.md:37]: I polished a bit the documentation and added volume coupling in the overview page of the adapter documentation.

@tirgendetwas is this formulation correct?

**Review by MakisH — APPROVED** [review]
The code looks quite clean, and this is very ready to merge! :rocket: :smile: Thanks, a lot!

I am running the tutorials and will merge once I have checked everything.

(The CI is just broken because the OpenFOAM packages are failing the integrity check. It works on my machine.)

File: docs/README.md
Comment (tirgendetwas) [reply docs/README.md:37]: I'm not sure, what do you mean by "the user needs to provide additional, application-specific code"? We haven't managed to read Pressure so far. Temperature should probably work if it's done through the enthalpy, but we haven't tried that

Comment (MakisH) [reply docs/README.md:37]: For temperature, I understand that one just needs to provided a codedSource that corresponds to their thermophysical model, no?

I can remove the reading part from temperature and pressure.

**Pull request merged**
